core/src/main/java/hudson/model/User.java

```diff
@@ -425,6 +440,20 @@ private LegitimateButUnknownUserDetails(String username) throws IllegalArgumentE
      */
     @Restricted(NoExternalUse.class)
     public @NonNull Authentication impersonate(@NonNull UserDetails userDetails) {
+        if (!ALLOW_IMPERSONATION_OF_DISABLED_USERS) {
+            if (!userDetails.isAccountNonLocked()) {
+                throw new UsernameNotFoundException(Messages.User_Locked(userDetails.getUsername()));
```

Comment:
Still trying to figure out in https://github.com/jenkinsci/ldap-plugin/pull/50#issuecomment-707216245 why this should be done in core rather than the plugin defining the `SecurityRealm`.

Reply:
Looking more closely at the code changes here, it seems like these exceptions could potentially be thrown in the LDAP `UserDetailsService` itself since that will be checked both by `User.getUserDetailsForImpersonation()` and therefore by `User.impersonate()` as well as the few places in Core that call the `impersonate(UserDetails)` variant (those call `getUserDetailsForImpersonation()` first). Being that the `impersonate(UserDetails)` method is restricted from external use, it seems safe to move this downward.

Reply:
> it seems like these exceptions could potentially be thrown in the LDAP `UserDetailsService` itself

Right, that was my suggestion. The core PR might still be useful hardening in case there were some unknown realm out there which actually did return an invalid `Authentication`, but we would not be relying on the core check.

Reply:
I've already reviewed all the `SecurityRealm` implementations for Jenkins in the wild (and proprietary ones) and found that none of them specify values other than `true` for these four properties. As I explained in greater detail in the LDAP plugin PR, it seems as though these properties are in the Spring Security API mostly for supporting [JDBC authentication providers](https://docs.spring.io/spring-security/site/docs/5.4.1/reference/html5/#servlet-authentication-jdbc) or other embedded user databases (e.g., Jenkins' own internal user database could optionally support these properties).